File: python/ch01_agent/agent_ch01/core/messages.py
"""聊天消息领域模型。

Java 对照：这里的 dataclass 就像 Java 的 record，专门保存数据。
消息不能随便用 dict，因为模型消息有固定角色，而且工具调用必须和工具结果一一配对。
"""

from dataclasses import dataclass
from typing import Literal

Role = Literal["system", "user", "assistant", "tool"]


class MessageContractError(Exception):
    """消息历史违反工具调用配对契约。

    把错误单独定义成一个类型，等价于 Java 中自定义业务异常，
    这样上层可以准确判断是消息格式错了，而不是网络错了。
    """


@dataclass(frozen=True, slots=True)
class ToolCall:
    """模型发出的“请程序帮我调用某个工具”的请求。"""
    id: str
    name: str
    arguments: str

    def __post_init__(self) -> None:
        _require_string(self.id, "tool call id")
        _require_string(self.name, "tool call name")
        _require_string(self.arguments, "tool call arguments", allow_empty=True)


@dataclass(frozen=True, slots=True)
class SystemMessage:
    """系统提示词：告诉模型它是谁、应该如何工作。"""
    role: Literal["system"]
    content: str


@dataclass(frozen=True, slots=True)
class UserMessage:
    """用户真正输入的问题。"""
    role: Literal["user"]
    content: str


@dataclass(frozen=True, slots=True)
class AssistantMessage:
    """模型的一次回答。

    `content` 有文本时表示模型在说话；`tool_calls` 非空时表示模型要程序做事。
    两者可以同时存在，但本章只关心是否存在工具调用。
    """
    role: Literal["assistant"]
    content: str | None
    tool_calls: tuple[ToolCall, ...] = ()

    def __post_init__(self) -> None:
        if self.content is not None:
            _require_string(self.content, "assistant content", allow_empty=True)
        ids = [call.id for call in self.tool_calls]
        if len(ids) != len(set(ids)):
            raise MessageContractError("assistant tool call ids must be unique")


@dataclass(frozen=True, slots=True)
class ToolMessage:
    """程序执行工具后的结果，必须带回原来的 tool_call_id。"""
    role: Literal["tool"]
    content: str
    tool_call_id: str


ChatMessage = SystemMessage | UserMessage | AssistantMessage | ToolMessage


def _require_string(value: object, field: str, allow_empty: bool = False) -> str:
    """统一检查字符串。

    Python 不像 Java 编译器那样能保证运行时输入一定是字符串，
    所以从模型或 JSON 进入系统时必须主动校验。
    """
    if not isinstance(value, str):
        raise MessageContractError(f"{field} must be a string")
    if not allow_empty and not value:
        raise MessageContractError(f"{field} must not be empty")
    return value


def tool_call(call_id: object, name: object, arguments: object) -> ToolCall:
    return ToolCall(
        _require_string(call_id, "tool call id"),
        _require_string(name, "tool call name"),
        _require_string(arguments, "tool call arguments", allow_empty=True),
    )


def system_message(content: str) -> SystemMessage:
    return SystemMessage("system", _require_string(content, "system content", True))


def user_message(content: str) -> UserMessage:
    return UserMessage("user", _require_string(content, "user content", True))


def assistant_message(content: str | None, tool_calls: tuple[ToolCall, ...] = ()) -> AssistantMessage:
    return AssistantMessage("assistant", content, tuple(tool_calls))


def tool_message(content: str, tool_call_id: str) -> ToolMessage:
    return ToolMessage(
        "tool",
        _require_string(content, "tool content", True),
        _require_string(tool_call_id, "tool_call_id"),
    )


def validate_tool_pairing(messages: list[ChatMessage] | tuple[ChatMessage, ...]) -> None:
    """确保每个 assistant 工具调用恰好有一个后续 tool 结果。

    例子：assistant 调用 call-1 后，历史中必须出现 tool_call_id=call-1 的结果。
    这一步相当于数据库保存前的关联完整性检查，避免把坏消息发给模型供应商。
    """
    pending: set[str] = set()
    for message in messages:
        if pending:
            if not isinstance(message, ToolMessage):
                raise MessageContractError(f"missing tool results for ids: {sorted(pending)!r}")
            if message.tool_call_id not in pending:
                raise MessageContractError(f"unexpected tool result id: {message.tool_call_id}")
            pending.remove(message.tool_call_id)
            continue
        if isinstance(message, ToolMessage):
            raise MessageContractError(f"orphan tool result id: {message.tool_call_id}")
        if isinstance(message, AssistantMessage):
            pending.update(call.id for call in message.tool_calls)
    if pending:
        raise MessageContractError(f"missing tool results for ids: {sorted(pending)!r}")

File: typescript/chapters/ch02/src/core/messages.ts
/**
 * 消息角色定义与不可变工厂函数。
 * 四种角色（system/user/assistant/tool）对应 Chat Completions API 的协议。
 * validateToolPairing 在发送请求前检查 tool_call_id 是否全部配对，
 * 防止 orphan 结果或缺项进入下一轮。工厂函数冻结消息对象以维持不可变性。
 */
export type Role = "system" | "user" | "assistant" | "tool";

// 消息契约阻止孤立 tool result 或未回填的模型工具调用进入下一轮。
export class MessageContractError extends Error {
  override readonly name = "MessageContractError";
}

export interface ToolCall {
  // 供应商生成的调用唯一 ID，tool result 通过它回填关联。
  readonly id: string;
  // 注册表查找使用的工具名称。
  readonly name: string;
  // 原始 JSON 参数文本，仅由 ToolRegistry 解析。
  readonly arguments: string;
}

export interface SystemMessage {
  // 判别标签；系统约束在每轮请求临时前置。
  readonly role: "system";
  // 系统提示正文。
  readonly content: string;
}

export interface UserMessage {
  // 判别标签；一次 run 的起点。
  readonly role: "user";
  // 用户原始任务文本。
  readonly content: string;
}

export interface AssistantMessage {
  // 判别标签；模型回复。
  readonly role: "assistant";
  // 最终文本或工具调用时的 null。
  readonly content: string | null;
  // 本回复提出且必须逐项回填的工具调用。
  readonly toolCalls: readonly ToolCall[];
}

export interface ToolMessage {
  // 判别标签；工具执行后的回填消息。
  readonly role: "tool";
  // 成功输出或带错误码的失败正文。
  readonly content: string;
  // 对应 assistant 调用的唯一 ID。
  readonly toolCallId: string;
}

// 四类消息的判别联合，供历史验证与 SDK 转换进行穷尽分支。
export type ChatMessage = SystemMessage | UserMessage | AssistantMessage | ToolMessage;

// 集中校验所有消息字段；allowEmpty 仅用于协议允许空正文的角色。
function requireString(value: unknown, field: string, allowEmpty = false): string {
  if (typeof value !== "string") {
    throw new MessageContractError(`${field} must be a string`);
  }
  if (!allowEmpty && value.length === 0) {
    throw new MessageContractError(`${field} must not be empty`);
  }
  return value;
}

// 创建并冻结工具调用，参数仅在注册表 prepare 阶段解析和校验。
export function toolCall(id: unknown, name: unknown, argumentsJson: unknown): ToolCall {
  return Object.freeze({
    id: requireString(id, "tool call id"),
    name: requireString(name, "tool call name"),
    arguments: requireString(argumentsJson, "tool call arguments", true),
  });
}

// 工厂函数冻结消息，确保模型请求的历史在创建后不被外部修改。
export function systemMessage(content: string): SystemMessage {
  return Object.freeze({ role: "system", content: requireString(content, "system content", true) });
}

// 创建用户消息；消息层允许空文本，不在此改变上层输入策略。
export function userMessage(content: string): UserMessage {
  return Object.freeze({ role: "user", content: requireString(content, "user content", true) });
}

// 创建模型消息并保证单一回复内的工具调用 ID 可无歧义配对。
export function assistantMessage(
  content: string | null,
  toolCalls: readonly ToolCall[] = [],
): AssistantMessage {
  if (content !== null) {
    requireString(content, "assistant content", true);
  }
  // 同一 assistant 消息不得重复声明 tool_call_id，否则无法无歧义回填结果。
  const ids = toolCalls.map((call) => call.id);
  if (new Set(ids).size !== ids.length) {
    throw new MessageContractError("assistant tool call ids must be unique");
  }
  return Object.freeze({ role: "assistant", content, toolCalls: Object.freeze([...toolCalls]) });
}

// 创建工具结果消息，要求关联 ID 非空以保持调用可审计。
export function toolMessage(content: string, toolCallId: string): ToolMessage {
  return Object.freeze({
    role: "tool",
    content: requireString(content, "tool content", true),
    toolCallId: requireString(toolCallId, "tool_call_id"),
  });
}

// 验证完整历史的调用/结果配对不变量，在请求供应商前阻止损坏状态外泄。
export function validateToolPairing(messages: readonly ChatMessage[]): void {
  // pending 保留当前 assistant 消息尚未收到结果的调用标识。
  const pending = new Set<string>();

  for (const message of messages) {
    if (pending.size > 0) {
      if (message.role !== "tool") {
        throw new MessageContractError(
          `missing tool results for ids: ${JSON.stringify([...pending].sort())}`,
        );
      }
      if (!pending.delete(message.toolCallId)) {
        throw new MessageContractError(`unexpected tool result id: ${message.toolCallId}`);
      }
      continue;
    }

    if (message.role === "tool") {
      throw new MessageContractError(`orphan tool result id: ${message.toolCallId}`);
    }
    if (message.role === "assistant" && message.toolCalls.length > 0) {
      for (const call of message.toolCalls) {
        pending.add(call.id);
      }
    }
  }

  if (pending.size > 0) {
    throw new MessageContractError(
      `missing tool results for ids: ${JSON.stringify([...pending].sort())}`,
    );
  }
}
